### ISSUE TITLE
Optimize collision model

### ISSUE THREAD

**Issue by dnk777** [Qfusion/qfusion#383]
I started working on it in my branch https://github.com/dnk777/qfusion/tree/cm-optimizations.

Its in very dirty state and is a subject of rebasing. To prevent losing comments due to forced updates please discuss it here.

I have some troubles loading Q3 BSP (collision does not work atm probably due to bad signbits or markbrush nums), but Q1 and Q2 code works well (movement and weapons feel right), and I test tracing optimizations on Q2 maps, since Q3 and Q2 maps use the same trace algorithm.

The ideas behind this work:
1) Optimize data layout to avoid pointer-chasing and cache misses.
2) Use prefetching of leaves for remaining tree-like structures while iterating over it. 
3) Try to avoid excessive branches that try to save few arithmetic ops.
4) Use GCC «goto computed label» extension https://gcc.gnu.org/onlinedocs/gcc/Labels-as-Values.html
5) Make recursive algorithms use a records stack instead.
6) Operate on FP sign bits instead of making FP branches.

Notes on what has been already implemented:

1)
* Storage of plane value inside nodes and brush sides, storage brush sides data immediately after the brush, storage of face brushes consequently to the face. This should aid hardware prefetcher while operating on these things.
* Contrary to the previous statement, split hot and cold data for markbrushes/markfaces. Avoid iterating over `brush **`, `face **`, iterate over a plain array of `markbrush`, `markplane` which contain important data for culling and a pointer to an actual brush/face. Accumulate feasible brush pointers while culling and prefetch their values before starting iterating over feasible brushes after culling.
* Add and use prefetch intrinsics for previous paragraph and for iterating over trees. Prefetch tree childs immediately after the current node address is known before any tests on it.

This sounds good but there is no real performance gain from it. First, amount of data for even large Q2 maps (the only maps available for testing atm) is very small, and since there is no really other code running aside from benchmark tracing, the data probably keeps dwelling in caches and never gets evicted. And also modern Intels have very good hardware prefetcher, so a manual prefetching can lead to performance improvement only for trees (and its used only for tree-like structures/access patterns, and needs to be tweaked yet).

3) From testing I can conclude that it is controversal and really depends of the actual environment in the code. 

4) This is a very nice feature, even if it has limited use in the current code structure. For example, it allows to ensure that switch would guaranteely use a jump table without any extra branches, and this table is located on stack (and thus is in cache). This feature has already proven to add speed up, and would allow to do much more if I figure out fast float bits extraction.   

5) Currently I have rewritten CM_RecursiveHullCheck() this way. The code seems to have an identical behavior (from testing by moving/shooting) but performance gain (if any) was not measured.

6) This was my major hope and turned to be a huge disappointment, because there is no way to get float bits as an integer with zero cost. Unions / pointer type casts force saving data on stack and back, and it is horribly slow, `__builtin_signbit` is also surprisingly slow (from what i have read in assembly it does some redundant ops, loads a 16-byte constant somewhere from memory, etc). I have tried GCC inline assembly that seem to produce sane output (like `movp %xmm0, %eax`) but looks like it kills optimizations. Actually if I try to avoid branches in this minor part of `CM_ClipBoxToBrush`

```
if( d2 > 0 ) {
	getout = true; // endpoint is not in solid
}
if( d1 > 0 ) {
	startout = true;
}
```

as 
```
uint32_t getout = 0, startout  = 0;
....

getout |= UFloatBits( -d2 ) & 0x80000000u;
startout |= UFloatBits( -d1 ) & 0x80000000u;
```

where `UFloatBits()` is crafted using any of the approach mentioned above, the code runs TWICE slower.

If I discover really fast float to int bits conversion, lots of code would benefit from it, including not only collision but `BoxOnPlaneSide`, etc. Converting sign bits and using these int values as an array index for dynamic goto labels would be very sweet.

Currently mine code runs a bit slower that the vanilla one. I use tracing from each item on a map to each other one as a benchmark. On map `q2ctf8.bsp` results are 22.5 ms and 19.5 ms consequently (around 6000 traces over the entire level).

**Comment by viciious** [comment]
Optimizing Q1 collisions is something that should be reaaaaally low priority anyway. As for the rest, I would suggest that you look into Bounding Interval Hierarchy, especially the implementation by LordHavoc, which can be found here: http://svn.icculus.org/twilight/trunk/darkplaces/?sortby=file#dirlist From what I've learned it outperforms BSP-based collision detection algorithms by considerable margin.

**Comment by dnk777** [comment]
I did not touch Q1 trace code for obvious reasons (nobody is going to run bots on these maps since there is no AAS compiler for Q1). I tested Q2 maps because these maps are only that might be loaded atm in the branch and they share the main trace algorithm. Porting a different collision system is not in my short-term plans. I think we could jam a substantial performance improvement from the existing (modified by me) code with relatively low effort  especially if I figure out cheap float to int bits conversion.

**Comment by viciious** [comment]
From my experience, switching to a better algorithm provides more substantial performance gains that anally optimizing your current one.
Evaluatting BIH perormance was already on my long-term TODO list anyway..

**Comment by dnk777** [comment]
Its hard to confess but mine attempts were pointless. Either the collision data already has quite good locality, or just hardware prefetchers/branch predictors on a modern Intel are so powerful, so there is no effect from prefetching / layout reordering / trying to eliminate branches. And also I have little knowledge of map formats, and probably due to it Q3BSP map support has been broken and I am not going to spend unpredictable amount of time fixing it. Maybe this attempt worth to be continued for some weak in-order or just old CPU (like Atom), so I'll keep the branch frozen as-is. I neither plan to work on CM in the current state anymore, nor plan to port something better for the CM. Optimizing CM was one of things I would like to be done as an aid for AI movement prediction.

**Comment by dnk777** [comment]
There are good news. Being jealous of the unfinished attempt, I decided to not give up, and map loading was fixed for Q3BSP/RBSP maps. Also a considerable speedup (up to 2x for huge maps like wbomb5) is possible. The code is not pushed yet, I need to spend some time cleaning it up.

**Comment by viciious** [comment]
Btw, I'm not sure what you mean by "Q3BSP map support being broken".

**Comment by dnk777** [comment]
> Btw, I'm not sure what you mean by "Q3BSP map support being broken".

I obviously meant that I have temporarily broken it in my branch doing the first attempt, it is fixed now.

Well I was again way too optimistic mainly based on poor performance of Clang-generated code for the upstream (gcc produces noticeably faster code having switched to it). I also was in hope on introduction of SIMD, but there is and should not be any gain from it for these scattered over the code math ops (i have tested SIMD in my private builds). 

I was relying on prefetching instructions too, but, first there is a hillarious HW bug (Ivy Bridge CPU can execute a SW-initiated prefetch once per 43 clocks, while all previous and next Intels can do two SW prefetch instructions per clock, it has been revealed as a well-known issue), and I spent a lot of time on such developer HW figuring out why prefetch has no effect, and there are still many opportunities to play with prefetch to probably eliminate way too high L1 miss rate in `CM_ClipBoxToBrush`).


If there is any gain, its very small, around few (3-5) percents. There are however cases when mine branch really performs better (on huge and/or terrain/patch based maps like old wooden wctf5, wbomb3, map simpsons_q3) where 10-20 percents of performance improvement has been achieved. 
 
I put this code in `/game/ai/ai.cpp` to test collision performance: https://gist.github.com/dnk777/b533a2b54d4fb5a440aa7e45c38f811a.  

I has an interest in MSVC results for huge maps like wbomb3, can you tell what's the difference with the upstream code?

Note that I am aware that currently there are redundant computations of patch data while loading.

**Comment by viciious** [comment]
Frankly, there's a good amount of changes on your branch that I don't like: too much cruft for too little gain. The 3-10% improvement probably comes from only one of the changes (probably from the cache friendliness) and if you could condense it into a small and reviewable patch, I might just as well merge it right away.

**Comment by dnk777** [comment]
The branch was in WIP state, so I have rebased it leaving only commits that contribute something to the final version. Also I have fixed duplicated patch data allocations/calculations while loading.

Could you tell me MSVC results for these tests (to be put in game/ai/ai.cpp)
https://gist.github.com/dnk777/3744e1fb34005e2aad38f77ec967dcd6
on such maps as wbomb2, wbomb5, wbomb3, wdm4 compared to the upstream? 
We have to ensure there is no performance regression for MSVC builds. The Bench1 results are move important for me (this test simulates `PMove()`-like type of load).

**Comment by viciious** [comment]
I'm not going to bother testing something I won't merge anyway, sorry.
All that prefetching nonsense is a no-go.

**Comment by dnk777** [comment]
What do you dislike, prefetch itself or its use here?

**Comment by viciious** [comment]
I dislike almost everything about your changes to the code.

**Comment by viciious** [comment]
Your changes add significant mostly unportable code clutter and hurt readability for no particular gain or performance improvement.

**Comment by dnk777** [comment]
1) Prefetch is a hint, its safe do do it on an illegal address. You can disable it building with `-DQF_NO_PREFETCH`.

2) Don't give up! I have pushed a commit that really boosts performance up to 2x, no jokes (measured @ wbomb5 using the test posted above). Its very dirty (I did not even apply code style fix on it), let me redo it tomorrow properly.

**Comment by viciious** [comment]
1. So what? I don't like the ugliness of code itself, not the compilation flags.
2. Now you're halfway through to BIH..

**Comment by dnk777** [comment]
1) Please explain what literally feels wrong. I am not stubborn to fix it and you are the owner, but I can't read thoughts.

2) I have an idea, split the world into a grid, build list of brushes in each cell and test the trace segment against all brushes in cells where the trace segment locates? Is not it something similar to BIH? Note that I very inexperienced in CM algorithms, I started this branch to just to speed up Trace_* stuff for AI purposes.

**Comment by viciious** [comment]
1. Like I said earlier, I don't like the micro-optimizations approach, in general.

2. I'm not an expert in BIH, LordHavoc is the guy you should be asking about BIH. What I did notice is that both approaches (yours and BIH) require decomposition of brushes into point windings.

**Comment by dnk777** [comment]
Cache-friendly memory layout and prefetching are not microoptimizations, rather macro :). I can drop some things that qualify as it, like rewriting CM_RecursiveHullCheck in iterative way.

**Comment by viciious** [comment]
Everything "under the 10%-improvement" falls into the micro-optimization category.
I don't mind the re-laying out of structures to improve the padding though.

**Comment by dnk777** [comment]
I have dropped two commits: 
1) The one that converts && -> & in several conditions;
2) Avoiding recursion in `CM_RecursiveHullCheck()`;

Note that these changes allowed 3-5 % of the performance gain.
The last commit that introduces reconstruction of brush bounds was cleaned up, and also a memory pool for temporary windings was added (this kind of improvement should be added for patch temporaries too).

The only change I plan is to add flat (non-hierarchical) grouping of brushes in a leaf for fast rejection culling of the entire group. Brushes of different groups are allowed to overlap (a brush may belong to many leafs, so there is already test for duplicate brush clipping). Several kinds of subdivisions might be tested, and the one with the highest score (lowest overlapping) should be chosen for a leaf. 

Also I still want to evaluate a SIMD version for `BoundsIntersect()` test.

I do not want to introduce anything more for reasons I have mentioned above (I'm inexperienced in it and do not want to spend much time on it, and there are quite satisfiable results in terms of performance improvement).

**Comment by viciious** [comment]
> 15:56 \<Vic\> hi
15:58 \<Vic\> I remember you mentioned the other day your BIH implementation is a marginally better performer than the BSP
15:58 \<Vic\> can you elaborate on that?
19:28 \<LadyHavoc\> two reasons for that - it's shallower (each leaf can hold up to N objects (brushes or triangles) which means less cache misses, I've found N can be from 8 to 32 with no significant difference in performance, if N is < 8 it starts performing worse), it has an early-out optimization (if a node has a gap between its child nodes, and the line segment falls in that gap, then the entire node can be culled)
19:28 \<LadyHavoc\> BSP always recurses to the bottom - which is useful if the leafs represent something meaningful (solid/non-solid) but an utterly useless feature as a culling structure
19:29 \<LadyHavoc\> note that you could turn BSP into a BIH style structure by just storing two extents (distances) along the plane instead of one, which enables the early-out behavior (but means that child nodes may overlap)
19:30 \<LadyHavoc\> BIH is also known for faster building (much faster than BSP) which makes it useful for realtime raytracers (which is what it was invented for)
19:31 \<LadyHavoc\> and BIH can be modified in place, which has also been done - sauerbraten uses BIH for mesh collisions on models, including skeletal models (where they update the extents of each node based on triangles moving around, but they don't change the node hierarchy or axes of splits)
19:32 \<LadyHavoc\> since for example a character model usually faces forward throughout its animation sets
19:32 \<LadyHavoc\> if it's animated, you'd want to make sure it follows a subdivision pattern like X->Y->Z->X->Y->Z->...
19:32 \<LadyHavoc\> so that the resulting node structure remains semi-optimal
19:33 \<LadyHavoc\> one other novelty with BIH (although this is purely a cute trick) is that when you get to the last few triangles in a region of the map, often they overlap in terms of AABB, to the point of even being identical AABB (two triangles of a floor quad) - BIH has no problem representing identical AABBs of children by just splitting on an arbitrary axis and storing the entire bounds as overlapping extents
19:34 \<LadyHavoc\> so technically BIH does not have to have any effective splits
19:34 \<LadyHavoc\> if you have a huge number of identical AABBs for children, it does not fail (but obviously those nodes won't do any useful culling)
19:34 \<LadyHavoc\> I mainly found that when using N = 1 (one triangle per leaf node)
19:35 \<LadyHavoc\> at N = 2 that case rarely happens
19:35 \<LadyHavoc\> but I recommend N = 8, 16 or even 32
19:35 \<LadyHavoc\> attempting a line trace (or even a brush vs brush trace) against 32 primitives turns out to be fast enough
19:36 \<LadyHavoc\> obviously the only reason for N = 32 would be if you want a very shallow tree to conserve memory
19:37 \<LadyHavoc\> it goes without saying that you could use a BIH builder as a way to create surfaces from a larger mesh, but I haven't fiddled with that
19:38 \<LadyHavoc\> actually wait, I have - I remember that in darkwar I load obj models that tend to be huge and high poly for terrain, and I just subdivided a surface in half along its best axis whenever it exceeded 100 triangles or so
19:38 \<LadyHavoc\> which is effectively the same algo that BIH uses
19:39 \<LadyHavoc\> a couple other performance tuning observations I learned from BIH:
19:39 \<LadyHavoc\> 1. don't try to compare AABBs of child nodes to recurse it as an AABB tree - comparing only one axis is much faster than comparing all 3
19:41 \<LadyHavoc\> 2. deciding the best axis for a split can be tricky - I use a scoring method that selects for which axis has the most even distribution of object centers on each side of the median
19:42 \<LadyHavoc\> using just an implicit X->Y->Z->X->Y->Z->... recursion pattern is totally valid, I haven't analyzed if it's better or not
19:42 \<LadyHavoc\> overall a BIH is just a much smarter and more flexible kdtree
19:43 \<LadyHavoc\> (kdtree often requires you to split the objects on arbitrary planes, or link them into nodes rather than leafs when they cross a plane, which results in far worse performance in practice)

**Comment by viciious** [comment]
Like I said, just don't bother with those micro-optimizations, throw them away and start from scratch. There's absolutely no chance I'm going to merge them into the master branch. Also, replacing integers with uint16_t is a rather pointless activity if static or stack variables are concerned.

This is the last time I'm repeating myself, don't waste your time on this.

**Comment by dnk777** [comment]
Please review this:

https://github.com/dnk777/qfusion/tree/cm-quick-optimizations

Only brush bounds reconstruction and testing is added. Its a minor change but there is still a substantial performance gain. If one decides to implement something non-BSP based for CM, this commit would also be useful, since it is the code where brush bounds are restored. I do not want to work on CM by myself, it is not a point of mine interest.

**Comment by viciious** [comment]
This is better but I think I'm going to bite the bullet and just port over LH's BIH implementation instead..

**Comment by viciious** [comment]
Btw, I find your changes a bit overwhelming as brush bounds can easily be reconstructed from 6 axial planes that always come first in the list of brushsides. See https://github.com/id-Software/Quake-III-Arena/blob/master/code/qcommon/cm_load.c#L215

**Comment by viciious** [comment]
Closing this issue for now, I'll open a new one for BIH when the time comes.
